### PR TITLE
ADM-727:[backend][docs]feat: Merge two API into one

### PR DESCRIPTION
**Pull request by Lei010**
## Summary

...

## Before

Generate Report divided into two apis.

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## After

Merge two API into one, and asyncly generate pipeline and code base metrics.

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## Note

_Null_

### CODE REVIEW

File: backend/src/test/java/heartbeat/service/report/GenerateReporterServiceTest.java [hunk 1]
```diff
@@ -1266,6 +1276,126 @@ void shouldPutReportInHandlerWhenCallSaveReporterInHandler() throws IOException
...
+
+	@Test
+	void shouldPutExceptionInHandlerWhenCallGenerateBoardReportThrowException()
+			throws IOException, InterruptedException {
+		ObjectMapper mapper = new ObjectMapper();
+		GenerateReportRequest reportRequest = mapper.readValue(new File(REQUEST_FILE_PATH),
+				GenerateReportRequest.class);
+		ReportResponse reportResponse = mapper.readValue(new File(RESPONSE_FILE_PATH), ReportResponse.class);
+		reportRequest.setCsvTimeStamp("1683734399999");
+		String boardTimeStamp = "board-1683734399999";
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/report/GenerateReporterServiceTest.java:1317]: what is `1683734399999`?
can we refactor with a readable name like `board-20240109232359`, even add milliseconds

must be changed with this comment

File: backend/src/test/java/heartbeat/service/report/GenerateReporterServiceTest.java [hunk 2]
```diff
@@ -1266,6 +1276,126 @@ void shouldPutReportInHandlerWhenCallSaveReporterInHandler() throws IOException
...
+	void shouldGeneratePipelineReportAndUpdatePipelineMetricsReadyWhenCallGeneratePipelineReport()
+			throws IOException, InterruptedException {
+		ObjectMapper mapper = new ObjectMapper();
+		GenerateReportRequest reportRequest = mapper.readValue(new File(REQUEST_FILE_PATH),
+				GenerateReportRequest.class);
+		reportRequest.setMetrics(List.of("Deployment frequency"));
+		ReportResponse reportResponse = mapper.readValue(new File(RESPONSE_FILE_PATH), ReportResponse.class);
+		GenerateReporterService spyGenerateReporterService = spy(generateReporterService);
+		reportRequest.setCsvTimeStamp("1683734399999");
+		String doraTimeStamp = "dora-1683734399999";
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/report/GenerateReporterServiceTest.java:1346]: same as `board`

File: backend/src/test/java/heartbeat/service/report/GenerateReporterServiceTest.java [hunk 3]
```diff
@@ -1266,6 +1276,126 @@ void shouldPutReportInHandlerWhenCallSaveReporterInHandler() throws IOException
...
+	void shouldGenerateCodebaseReportAndUpdateCodebaseMetricsReadyWhenCallGeneratePipelineReport()
+			throws IOException, InterruptedException {
+		ObjectMapper mapper = new ObjectMapper();
+		GenerateReportRequest reportRequest = mapper.readValue(new File(REQUEST_FILE_PATH),
+				GenerateReportRequest.class);
+		reportRequest.setMetrics(List.of("Lead time for changes"));
+		ReportResponse reportResponse = mapper.readValue(new File(RESPONSE_FILE_PATH), ReportResponse.class);
+		GenerateReporterService spyGenerateReporterService = spy(generateReporterService);
+		reportRequest.setCsvTimeStamp("1683734399999");
+		String codebaseTimeStamp = "github-1683734399999";
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/report/GenerateReporterServiceTest.java:1377]: same as board

**Review by guzhongren — COMMENTED** [review]
must be Changed with the comments

File: backend/src/test/java/heartbeat/service/report/GenerateReporterServiceTest.java (hunk 1 above)
Comment (Lei010) [reply backend/src/test/java/heartbeat/service/report/GenerateReporterServiceTest.java:1317]: Done

File: backend/src/test/java/heartbeat/service/report/GenerateReporterServiceTest.java (hunk 2 above)
Comment (Lei010) [reply backend/src/test/java/heartbeat/service/report/GenerateReporterServiceTest.java:1346]: Done

File: backend/src/test/java/heartbeat/service/report/GenerateReporterServiceTest.java (hunk 3 above)
Comment (Lei010) [reply backend/src/test/java/heartbeat/service/report/GenerateReporterServiceTest.java:1377]: Done

File: backend/src/main/java/heartbeat/service/report/GenerateReporterService.java [hunk 4]
```diff
@@ -213,17 +201,81 @@ else if (jsonObject.has("key")) {
...
 	}
 
+	public void generateBoardReport(GenerateReportRequest request) {
+		initializeMetricsDataReadyInHandler(request.getCsvTimeStamp(), request.getMetrics());
+
+		CompletableFuture.runAsync(() -> {
+			try {
+				saveReporterInHandler(generateReporter(request), IdUtil.getBoardReportId(request.getCsvTimeStamp()));
+				updateMetricsDataReadyInHandler(request.getCsvTimeStamp(), request.getMetrics());
+				log.info(
```
[the remark this answers is not in the document]
Comment (Lei010) [reply backend/src/main/java/heartbeat/service/report/GenerateReporterService.java:211]: Done

File: backend/src/main/java/heartbeat/service/report/GenerateReporterService.java [hunk 5]
```diff
@@ -213,17 +201,81 @@ else if (jsonObject.has("key")) {
...
+
+	private void generatePipelineReport(GenerateReportRequest request) {
+		CompletableFuture.runAsync(() -> {
+			try {
+				GenerateReportRequest pipelineRequest = request.convertToPipelineRequest(request);
+				saveReporterInHandler(generateReporter(pipelineRequest),
+						IdUtil.getDoraReportId(pipelineRequest.getCsvTimeStamp()));
+				updateMetricsDataReadyInHandler(pipelineRequest.getCsvTimeStamp(), pipelineRequest.getMetrics());
+				log.info(
+						"Successfully generate pipeline report, _metrics: {}, _considerHoliday: {}, _startTime: {}, _endTime: {}, _pipelineReportId: {}",
```
[the remark this answers is not in the document]
Comment (Lei010) [reply backend/src/main/java/heartbeat/service/report/GenerateReporterService.java:243]: Done

File: backend/src/main/java/heartbeat/service/report/GenerateReporterService.java [hunk 6]
```diff
@@ -213,17 +201,81 @@ else if (jsonObject.has("key")) {
...
+
+	private void generateCodeBaseReport(GenerateReportRequest request) {
+		CompletableFuture.runAsync(() -> {
+			try {
+				GenerateReportRequest codebaseRequest = request.convertToCodeBaseRequest(request);
+				saveReporterInHandler(generateReporter(codebaseRequest),
+						IdUtil.getCodeBaseReportId(codebaseRequest.getCsvTimeStamp()));
+				updateMetricsDataReadyInHandler(codebaseRequest.getCsvTimeStamp(), codebaseRequest.getMetrics());
+				log.info(
+						"Successfully generate codebase report, _metrics: {}, _considerHoliday: {}, _startTime: {}, _endTime: {}, _codeBaseReportId: {}",
```
[the remark this answers is not in the document]
Comment (Lei010) [reply backend/src/main/java/heartbeat/service/report/GenerateReporterService.java:262]: Done

File: backend/src/test/java/heartbeat/controller/report/GenerateReporterControllerTest.java [hunk 7]
```diff
@@ -161,7 +161,7 @@ void shouldReturnAcceptedStatusAndCallbackUrlAndIntervalWhenCallBoardReports() t
 		doNothing().when(generateReporterService).updateMetricsDataReadyInHandler(any(), any());
 
 		MockHttpServletResponse response = mockMvc
-			.perform(post("/reports/board").contentType(MediaType.APPLICATION_JSON)
+			.perform(post("/reports/v1/board").contentType(MediaType.APPLICATION_JSON)
```
[the remark this answers is not in the document]
Comment (Lei010) [reply backend/src/test/java/heartbeat/controller/report/GenerateReporterControllerTest.java]: Done

File: backend/src/main/java/heartbeat/controller/report/GenerateReportController.java [hunk 8]
```diff
@@ -62,7 +65,7 @@ public ResponseEntity<ReportResponse> generateReport(@PathVariable String report
 		return ResponseEntity.status(HttpStatus.OK).body(reportResponse);
 	}
 
-	@PostMapping("/board")
+	@PostMapping("/v1/board")
```
[the remark this answers is not in the document]
Comment (Lei010) [reply backend/src/main/java/heartbeat/controller/report/GenerateReportController.java]: Done

**Review by mikeyangyun — APPROVED** [review]
LGTM

**Pull request merged**